### DISCUSSION TITLE
Squash & Merge Implementation

### DISCUSSION THREAD

**Discussion by VadimBelov** [community/community#22860] [New to GitHub]
<p>Hello, </p>
<p>I’m trying to manually implement Github’s Squash&amp;Merge<span> </span>functionality.</p>
<p>After looking many forums like <a href="http://stackoverflow.com/" rel="nofollow noopener">stackoverflow.com</a> I’m still struggling with the issue.</p>
<p>The scenario is as follows:</p>
<ul>
<li>master is a protected branch (configured on github)</li>
<li>a service user _svc _is permitted to push to master (configured on github)</li>
<li>other users don’t have such permissions</li>
<li>github PR status check is set to SUCCESS before execution of any set of git commands below</li>
</ul>
<ol>
<li>This merges successfully without squash:<br>
git checkout origin/master</li>
</ol>
<p>git merge ${PR-Branch}<br>
git push origin HEAD:master<br>
git push origin --delete ${PR-Branch} </p>
<ol start="2">
<li>This closes the PR, but there is no update seen on master:</li>
</ol>
<p>git checkout origin/master<br>
git merge --squash --commit ${PR-Branch}<br>
git push origin HEAD:master<br>
git push origin --delete ${PR-Branch}</p>
<ol start="3">
<li>This fails to push to master with the error “GH006: Protected branch update failed”  (despite that the PR is set to SUCCESS):</li>
</ol>
<p>git checkout origin/master  <br>git merge --squash ${PR-Branch}<br>
git commit -am"comment"<br>
git push origin HEAD:${m_mainBranch}<br>
git push origin --delete ${m_prBranch}</p>
<p>Will appreciate your help on this.</p>
<p>Thanks,<br>
Vadim </p>

**Comment by lee-dohm** [comment]
<p>Does it work if you push from a local branch named <code>master</code> to the remote branch <code>master</code> instead of using a detached HEAD like in your examples?</p>

**Comment by VadimBelov** [comment]
<p><a class="mention" href="/lee-dohm">@lee-dohm</a>, thanks for your response.</p>
<p>I’ve tried this:</p>
<pre><code>git checkout master  
git merge --squash ${PR-Branch}  
git commit --file .git/SQUASH\_MSG   
git push -u origin master  
git push origin --delete ${PR-Branch}
</code></pre>
<p>The first push still results in:</p>
<pre><code>EXITCODE 0remote: error: GH006: Protected branch update failed for refs/heads/master
</code></pre>

**Comment by lee-dohm** [comment]
<p>Are you doing the push using the credentials for the <em>svc</em> user?</p>

**Comment by lee-dohm** [comment]
<p>I’ve consulted with some others on what you’re running into. We’ve got a few questions:</p>
<ol>
<li>Is this repo that you’re experimenting with private?</li>
<li>If so, please contact private Support at <a href="https://github.com/contact" rel="nofollow noopener">https://github.com/contact</a> with all of this information so they can take a deeper look at things</li>
<li>If not, please let us know what repo you’re accessing to get these errors</li>
<li>What software, if any beyond plain old Git, are you using to perform these actions?</li>
<li>If it is only Git, which version of Git is it?</li>
</ol>

**Comment by VadimBelov** [comment]
<p>Yes,<br>
Prior to the git commands I’ve described previously we run those:</p>
<pre><code>git config user.name ${GIT_USERNAME}  
git config user.email "jenkins@jenkins.com"  
git remote set-url origin https://${GIT_USERNAME}:${ctx.GIT_PASSWORD}@${GIT_REPO}
</code></pre>
<p>It is the same <em>svc</em> user we use to checkout the repository in the Jenkins job.</p>

**Comment by VadimBelov** [comment]
<ol>
<li>
<p>It is a private repo indeed, will try to contact the private support as you suggested</p>
</li>
<li>
<p>It is running in a shell script section as the last stage of a Jenkins pipeline</p>
</li>
<li>
<p>git version on Jenkins is 2.11.0</p>
</li>
</ol>

**Comment by VadimBelov** [comment, marked as answer]
<p>Actually this is solvable using github API.</p>
<pre><code>curl -X PUT -H "Authorization: token ${adminToken}" --data '{"merge_method":"squash"}' https://github.&lt;&gt;.com/api/v3/repos/:org/:repo/pulls/${PR_ID}/merge
</code></pre>
